Score CTC prefix beams with KenLM

reuben opened the pull request:
Opening a PR to test the TaskCluster setup.

reuben commented:
FWIW, all the tests passed: https://tools.taskcluster.net/groups/Vxu6YJ_vR-6GFI-vi0Z0nA

I need to figure out a solution for users in general, probably downloading the appropriate native_client.tar.xz automatically from the bin/run-* scripts and extracting the library so that training can be done without having to set up a TensorFlow build environment.

reuben commented:
@kdavis-mozilla @lissyx I've split the changes into logical chunks where possible. I'll work on the scripts I mentioned above before merging the PR, but the commits that are here are ready to be reviewed.

lissyx commented on native_client/BUILD:
```diff
@@ -34,3 +34,35 @@ cc_library(
+    name = "ctc_decoder_with_kenlm",
+    srcs = ["beam_search.cc",
+            "alphabet.h",
+            "trie_node.h"] +
```
nit: the "] +" should be on the next line

lissyx commented on native_client/BUILD:
```diff
@@ -34,3 +34,35 @@ cc_library(
+        "trie_node.h",
+        "alphabet.h",
+    ] + glob(["kenlm/lm/*.cc", "kenlm/util/*.cc", "kenlm/util/double-conversion/*.cc",
+                 "kenlm/lm/*.hh", "kenlm/util/*.hh", "kenlm/util/double-conversion/*.h"],
```
nit: alignment

lissyx commented on native_client/BUILD:
```diff
@@ -34,3 +34,35 @@ cc_library(
+                 "kenlm/lm/*.hh", "kenlm/util/*.hh", "kenlm/util/double-conversion/*.h"],
+                exclude = ["kenlm/*/*test.cc", "kenlm/*/*main.cc"]),
+    includes = ["kenlm"],
+    copts = ['-std=c++11'],
```
nit: use " instead of '

lissyx commented on native_client/beam_search.cc:
```diff
@@ -0,0 +1,546 @@
+limitations under the License.
+==============================================================================*/
+
+// This test illustrates how to make use of the CTCBeamSearchDecoder using a
```
I think those are not needed anymore :)

lissyx commented on native_client/beam_search.cc:
```diff
@@ -0,0 +1,546 @@
+      c->set_output(out_idx++, c->Matrix(batch_size, top_paths));
+      return tf::Status::OK();
+    })
+    .Doc(R"doc(
```
I think the doc needs to be updated to include the new parameters

reuben replied:
I didn't change this line so GitHub is still showing the comment but it's been fixed.

lissyx commented on native_client/beam_search.cc:
```diff
@@ -0,0 +1,546 @@
+    OP_REQUIRES_OK(ctx, ctx->GetAttr("top_paths", &top_paths));
+    decode_helper_.SetTopPaths(top_paths);
+
+    // const tf::Tensor* model_tensor;
```
Is that leftover from previous testing ?

kdavis-mozilla replied:
Could you remove this commented  out code?

lissyx commented on native_client/generate_trie.cpp:
```diff
@@ -0,0 +1,66 @@
+}
+
+int main(void) {
+  return generate_trie("/Users/remorais/Development/DeepSpeech/data/alphabet.txt",
```
Seems to be leftover as well :/

lissyx commented on native_client/generate_trie.cpp:
```diff
@@ -0,0 +1,66 @@
+  ifs.open(vocab_path, std::ifstream::in);
+
+  if (!ifs.is_open()) {
+    std::cout << "unable to open vocabulary" << std::endl;
```
errors should go to stderr

lissyx commented on .tc.training.yml:
```diff
@@ -11,16 +11,17 @@ payload:
       echo -e "#!/bin/bash\nset -xe\nexport PATH=/home/build-user/bin:$PATH && env && id && wget https://github.com/git-lfs/git-lfs/releases/download/v2.2.1/git-lfs-linux-amd64-2.2.1.tar.gz -O - | tar -C /tmp -zxf - && PREFIX=/home/build-user/ /tmp/git-lfs-2.2.1/install.sh && mkdir ~/DeepSpeech/ && git clone --quiet {{ GITHUB_HEAD_REPO_URL }} ~/DeepSpeech/ds/ && cd ~/DeepSpeech/ds && git checkout ...
       {{ SYSTEM_DO_CLONE }} &&
-      sudo -H -u build-user TENSORFLOW_WHEEL=${TENSORFLOW_WHEEL} /bin/bash /home/build-user/DeepSpeech/ds/tc-train-tests.sh 2.7.13
+      sudo -H -u build-user TENSORFLOW_WHEEL=${TENSORFLOW_WHEEL} {{ TASK_ENV_VARS }} /bin/bash /home/build-user/DeepSpeech/ds/tc-train-tests.sh 2.7.13
```
We don't need `{{ TASK_ENV_VARS }}`, just `{{ DEEPSPEECH_ARTIFACTS_ROOT }}`

lissyx commented on DeepSpeech.py:
```diff
@@ -484,7 +538,7 @@ def calculate_mean_edit_distance_and_loss(model_feeder, tower, dropout):
 
     # Beam search decode the batch
-    decoded, _ = tf.nn.ctc_beam_search_decoder(logits, batch_seq_len, merge_repeated=False)
+    decoded, _ = decode_with_lm(logits, batch_seq_len, merge_repeated=False, beam_width=1024)
```
Would it make sense to be able to keep the ability to use the vanilla tensorflow decoder ?

reuben replied:
I don't think anyone will want to use it, and I'd rather avoid forking the execution/maintenance paths and have some part of the code break because it's not being tested.

lissyx commented on native_client/generate_trie.cpp:
```diff
@@ -0,0 +1,66 @@
+  }
+
+  std::ofstream ofs;
+  ofs.open(trie_path);
```
shouldn't we check that as well? if we cannot open the output file

lissyx commented on native_client/generate_trie.cpp:
```diff
@@ -0,0 +1,66 @@
+  ofs.open(trie_path);
+
+  std::string word;
+  while (ifs >> word) {
```
So, this is going to read line by line the worlds.txt file?

reuben replied:
Word by word, it splits on whitespace.

lissyx commented on native_client/trie_node.h:
```diff
@@ -0,0 +1,125 @@
+    }
+  }
+
+  int GetFrequency() {
```
That is a bit misleading, a frequency is not exactly a number of occurences in my mind, or am I missing something ?

reuben replied:
Sure, it's more of a count. I'll rename it (or delete it, since there's no users).

lissyx commented on native_client/generate_trie.cpp:
```diff
@@ -0,0 +1,66 @@
+    for_each(word.begin(), word.end(), [](char& a) { a = tolower(a); });
+    lm::WordIndex vocab = GetWordIndex(model, word);
+    float unigram_score = ScoreWord(model, vocab);
+    root.Insert(word.c_str(), [&a](char c) {
```
nit: For readability, I guess the translator should be completely on a new line:
```
root.Insert(
    word.c_str(),
    [&a](char c) {

    },
    vocab,
    unigram_score);
```

lissyx commented on native_client/trie_node.h:
```diff
@@ -0,0 +1,125 @@
+      TrieNode *child = children[vocabIndex];
+      if (child == nullptr)
+        child = children[vocabIndex] = new TrieNode(vocab_size);
+      child->Insert(word + 1, translator, lm_word, unigram_score);
```
Are we playing with pointers here with the `+ 1` ?

reuben replied:
Yes.

lissyx commented on native_client/trie_node.h:
```diff
@@ -0,0 +1,125 @@
+    if (wordCharacter != '\0') {
+      int vocabIndex = translator(wordCharacter);
+      TrieNode *child = children[vocabIndex];
+      if (child == nullptr)
```
nit: i'm not a big fan of brace-less conditions, it's often root for nasty bugs in the future :)

lissyx commented on native_client/trie_node.h:
```diff
@@ -0,0 +1,125 @@
+  }
+
+  static void ReadFromStream(std::istream& is, TrieNode* &obj, int vocab_size) {
+    int prefixCount;
```
having a local variable of a static class named the same way a member of the class is confusing, imho

reuben replied:
I renamed member variables to have a _ suffix.

lissyx commented on native_client/generate_trie.cpp:
```diff
@@ -0,0 +1,66 @@
+  Model::State in_state = model.NullContextState();
+  Model::State out;
+  lm::FullScoreReturn full_score_return;
+  full_score_return = model.FullScore(in_state, vocab, out);
```
Why is there this `out` that we don't use ?

lissyx replied:
Asking because according to that comment https://github.com/mozilla/DeepSpeech/pull/805/commits/f1e859ef7b94ec01dcea67d91b7dfe7d271fe80c#diff-7a0b3b152c06fc03b20e63b72c26583aR11 (which I don't completely get the point, so I might just be misunderstanding something there), it seems like `out_state` has some use.

reuben replied:
Keeping state is a performance optimization when scoring a sentence of various words/n-grams, in this function we want to score the word independently of any context so that's why we throw it away.

lissyx replied:
Thanks, maybe worth a comment then ?

lissyx commented on native_client/beam_search.cc:
```diff
@@ -0,0 +1,546 @@
+};
+
+// CTC beam search
+class CTCBeamSearchDecoderOp : public tf::OpKernel {
```
That class name is already used by Tensorflow vanilla CTC Beam Search decoder. Can we rename to something matching the op? Like `CTCBeamSearchDecoderWithLM`.

reuben replied:
Technically this is `::CTCBeamSearchDecoderOp` rather than `::tensorflow::CTCBeamSearchDecoderOp`, but you're right, I'll rename it :)

kdavis-mozilla commented on native_client/beam_search.cc:
```diff
@@ -0,0 +1,546 @@
+  }
+
+  void Compute(tf::OpKernelContext *ctx) override {
+    const tf::Tensor *inputs;
```
Why such C based C++? Introducing all the variables at the start of the function.

reuben replied:
This code is copied from `CTCBeamSearchDecoder` in tensorflow, mostly unmodified. All I change is which beam scorer to use. I can make the code more C++-y, but then if we ever need to uplift changes from there to here it's harder.

kdavis-mozilla replied:
OK, understood.

lissyx commented on native_client/trie_node.h:
```diff
@@ -0,0 +1,125 @@
+    }
+    if (wordCharacter != '\0') {
+      int vocabIndex = translator(wordCharacter);
+      TrieNode *child = children[vocabIndex];
```
Agree, and with braces it would be much cleaner and future proof :)

reuben commented on native_client/kenlm/LICENSE:
```diff
@@ -0,0 +1,25 @@
+have their own licenses, listed below.  See comments in those files for more
+details.  
+
+util/getopt.* is getopt for Windows
```
We don't need them, so I'll remove them.

reuben commented on native_client/generate_trie.cpp:
```diff
@@ -0,0 +1,66 @@
+#include "trie_node.h"
+#include "alphabet.h"
+
+typedef lm::ngram::ProbingModel Model;
```
The ProbingModel choice is probably not the best for Pipsqueak, but changing this in the future would require very little effort, just switching the typedef, so I think it's OK to start with it in the Deep Speech context so we don't waste time on our experiments.

reuben commented on native_client/generate_trie.cpp:
```diff
@@ -0,0 +1,66 @@
+
+  std::string word;
+  while (ifs >> word) {
+    for_each(word.begin(), word.end(), [](char& a) { a = tolower(a); });
```
I think a better approach is to just remove this code and not make any assumptions about the vocabulary file. Maybe the user wants to use all upper case characters instead of lower case, maybe their language doesn't even have such a concept. It should be up to the user to make sure their vocabulary file matches their alphabet.txt file.

reuben commented on native_client/generate_trie.cpp:
```diff
@@ -0,0 +1,66 @@
+int main(void) {
+  return generate_trie("/Users/remorais/Development/DeepSpeech/data/alphabet.txt",
+                       "/Users/remorais/Development/DeepSpeech/data/lm/lm.binary",
+                       "/Users/remorais/Development/DeepSpeech/data/lm/vocab.txt",
```
`vocab.txt` is a text file with space separated words, not necessarily one per line. I've used the vocabulary file published by LibriSpeech here, but we should use the same vocabulary that was used to build the language model itself. I'll fix that.

lissyx commented on tc-tests-utils.sh:
```diff
@@ -47,7 +47,7 @@ assert_correct_ldc93s1()
 }
 
-download_material()
+download_native_client_files()
```
I am stealing part of that because I need to perform something similar in https://github.com/mozilla/DeepSpeech/pull/810

kdavis-mozilla commented:
@reuben I guess you've seen [Bug 1396777](https://bugzilla.mozilla.org/show_bug.cgi?id=1396777). 

Do you want to fold those changes in to this PR or do a separate PR?

reuben commented:
I'll do them here. I have an additional question for the bug before I do that, though.

reuben commented on native_client/beam_search.cc:
```diff
@@ -0,0 +1,546 @@
+  float delta_score;
+  std::string incomplete_word;
+  TrieNode *incomplete_word_trie_node;
+  lm::ngram::ProbingModel::State model_state;
```
It's a good choice for our current use case (fast at the expense of memory). Switching this out in the future is simple.

reuben commented on native_client/beam_search.cc:
```diff
@@ -0,0 +1,546 @@
+      // TODO try two options
+      // 1) unigram score added up to language model scare
+      // 2) langugage model score of (preceding_words + unigram_word)
+      to_state->score = min_unigram_score + to_state->language_model_score;
```
They do not. Until there's a space we accumulate min unigram scores and then when a word boundary is reached we score it with the LM and update language_model_score.

reuben commented:
@lissyx the TC build is failing because the version of GCC/libstdc++ on the builders does not support `codecvt`. How would I go about upgrading to GCC 5.0/the corresponding libstdc++ version?

Alternatively, if that's too complicated, I can use Boost.Locale, but that's a new dependency I'd rather avoid.

reuben commented:
@kdavis-mozilla I've addressed all review comments. Could you take a second look? In particular, the trie_node.h, beam_search.cc and DeepSpeech.py changes. Thanks!

lissyx commented:
@reuben You might be able to get `#include <codecvt>` by installting GCC 5, however:
 - this will require that we also build Tensorflow with that
 - this needs to be checked on OSX as well
 - ARM RPi3 toolchain is GCC 4.9

There is nothing newer available for RPi3, switching toolchain for ARM cross compilation might be very tricky.

reuben commented:
@lissyx that's what I was afraid of. I remember having trouble building TensorFlow with GCC>4.9, so I think the easier solution is to go with Boost.Locale.

kdavis-mozilla commented:
@reuben On the review. Should I wait until the switch to Boost.Locale?

reuben commented:
@kdavis-mozilla I'll try to upgrade to GCC 5.0 first. But yeah, I'll let you know once it's ready for review.

reuben commented:
There's no way to pass the appropriate include path to Bazel from the external world, so I vendored the necessary headers instead in native_client/boost_locale.

reuben commented on native_client/beam_search.cc:
```diff
@@ -219,36 +222,43 @@ class KenLMBeamScorer : public tf::ctc::BaseBeamScorer<KenLMBeamState> {
  private:
-  Model *model;
-  Alphabet *alphabet;
+  Model model;
```
Oversight! Will fix it.

reuben commented on DeepSpeech.py:
```diff
@@ -142,6 +142,12 @@
+tf.app.flags.DEFINE_string  ('lm_binary_path',       'data/lm/lm.binary', 'path to the language model binary file created with KenLM')
+tf.app.flags.DEFINE_string  ('lm_trie_path',         'data/lm/trie', 'path to the language model trie file created with native_client/generate_trie')
+tf.app.flags.DEFINE_integer ('beam_width',        1024,       'beam width used in the CTC decoder when building candidate transcriptions')
+tf.app.flags.DEFINE_float   ('lm_weight',         2.15,        'the alpha hyperparameter of the CTC decoder. Language Model weight.')
```
I don't remember and can't find where I got the beta' terminology from. I'll reword the explanation.
